ledger/streaming/client/src/ledgerdb_sink.rs

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2018-2021 The MobileCoin Foundation
+
+//! Helper class for creating a blocksink
+
+use std::sync::{Arc, RwLock};
+
+use futures::stream::{Stream, StreamExt};
+
+use mc_common::logger::{log, Logger};
+use mc_ledger_db::Ledger;
+use mc_ledger_streaming_api::StreamResult;
+use mc_transaction_core::BlockData;
+
+use crate::streaming_futures::WriteBlock;
+
+/// A block sink that takes blocks from a passed stream and puts them into
+/// ledger db. This sink should live downstream from a verification source that
+/// has already done block content, scp, and avr validation and thus is
+/// considered a trusted stream
+pub struct BlockSink<
+    'a,
+    L: Ledger + 'a,
+    F: Fn<US>,
+    US: Stream<Item = StreamResult<BlockData>>,
+    BS: Stream<Item = WriteBlock<'a, L>>,
+> {
+    /// Upstream block data
+    upstream: US,
+
+    /// Stream of block data
+    block_sink: BS,
+
+    /// Logger running
+    logger: Logger,
+}
+
+impl<
+    'a,
+    L: Ledger + 'a,
+    F: Fn(US) -> WriteBlock<'a, L>,
+    US: Stream<Item = StreamResult<BlockData>>,
+    BS: Stream<Item = F> + std::marker::Unpin,
+> BlockSink<'a, L, F, US, BS> {
+    /// Initialize a sink from an upstream source
+    pub fn new(upstream: US, logger: Logger, ledger: Arc<RwLock<&'a mut L>>, buffer: usize) -> Self {
+        let block_sink = Self::sink_from_upstream(upstream.clone(), ledger.clone(), buffer);
+        Self {
+            upstream,
+            block_sink,
+            logger,
+        }
+    }
+
+    pub fn sink_from_upstream(
+        upstream: US,
+        ledger: Arc<RwLock<&'a mut L>>,
+        buffer: usize
+    ) -> BS {
+        upstream.
+            map( move |block_data| { WriteBlock::new(ledger ,block_data.unwrap()) })
+            .buffered(buffer)
+    }
+
+    /// Execute storing the blocks in an opinionated way
+    pub async fn store_blocks(&mut self, limit: Option<usize>) {
+        let mut counter = 0usize;
+        while let Some(index) = self.block_sink.next() {
+            log::trace!(self.logger, "successfully wrote block {}", index);
+            if let Some(limit) = limit {
+                counter += 1;
+                if counter > limit  { break }
```

Comment:
IIUC we want >= here.
```suggestion
                if counter >= limit  { break }
```